### PR TITLE
FusionCache.CompleteBackgroundFactory => Using async/await instead of ".Result"

### PR DESCRIPTION
**Pull request by henriqueholtz**
According to [.NET Async guide](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md), it's better to use `.GetAwaiter().GetResult()` than `.Result`.

EDIT: I haven't noticed it was possible in this case earlier but as the code-reviewer suggested, I edited to use async/await instead.

### CODE REVIEW

File: src/ZiggyCreatures.FusionCache/FusionCache.cs [hunk 1]
```diff
@@ -400,7 +400,7 @@ private void CompleteBackgroundFactory<TValue>(string operationId, string key, F
 					options.ReThrowBackplaneExceptions = false;
 
 					// ADAPTIVE CACHING UPDATE
-					var lateEntry = FusionCacheMemoryEntry<TValue>.CreateFromOptions(antecedent.Result, options, false, ctx.LastModified, ctx.ETag, null);
+					var lateEntry = FusionCacheMemoryEntry<TValue>.CreateFromOptions(antecedent.GetAwaiter().GetResult(), options, false, ctx.LastModified, ctx.ETag, null);
```
Comment (gcsuk) [inline src/ZiggyCreatures.FusionCache/FusionCache.cs:403]: Whilst anything is better than .Result, GetAwaiter().GetResult() is still making things synchronous.

Could it not just be awaited?

**Comment by jodydonetti** [comment]
Hi all, I'll comment more later (sorry, super busy with my daily job) but this is a non-issue: some lines above I've already verified that the task is completed, so this will not block.

File: src/ZiggyCreatures.FusionCache/FusionCache.cs (hunk 1 above)
Comment (henriqueholtz) [reply src/ZiggyCreatures.FusionCache/FusionCache.cs]: Good point, I haven't noticed that! Done.

**Comment by henriqueholtz** [comment]
> Hi all, I'll comment more later (sorry, super busy with my daily job) but this is a non-issue: some lines above I've already verified that the task is completed, so this will not block.
> 
> If I haven't missed something it's like [here](https://github.com/mgravell/DistributedCacheDemo/blob/main/DistributedCacheExtensions.cs#L58) (by Marc Gravell).

Is it enough to not to migrate to async/await?

I don't think so. According to [.NET Async guide](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md), a task which is usually already completed should be threated differently (using `ValueTask` for example).

**Comment by jodydonetti** [comment]
Hi @henriqueholtz , sorry for the delay.

> I don't think so. According to [.NET Async guide](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md), a task which is usually already completed should be threated differently (using `ValueTask` for example).

The task in question is not usually already completed, it's the (user provided) factory to grab the fresh version of the data, and it can go from being instantaneous to taking seconds or minutes, and we don't have control over that. I'm checking if, at that moment, is already completed but that does not mean it usually is.

Also, `ValueTask` is preferable over `Task` to avoid allocations, but it does not have anything to do with it being usually completed, and it's a task because that is the signature for the factory which cannot be changed now (it's a breaking change).

Having said that I see that the latest version of the PR is using async code directly: I'd like to check it better and do some tests, it may in fact work better.

Let me check it and will come back with something soon.

Thanks!

**Review by jodydonetti — CHANGES_REQUESTED** [review]
An additional `await` means an additional async state machine and so on, which in this case (where we already know the task is completed) it's a waste.

Let's get back to `GetAwaiter().GetResult()` and Ithink we can merge this.

**Comment by henriqueholtz** [comment]
> An additional `await` means an additional async state machine and so on, which in this case (where we already know the task is completed) it's a waste.
> 
> Let's get back to `GetAwaiter().GetResult()` and I think we can merge this.

[Disagree and commit] I tend to disagree about using `GetAwaiter().GetResult()` instead of `async/await` + `ValueTask` (maybe it represent a gap in my knowledge). Even though it's done!

**Comment by jodydonetti** [comment]
> [Disagree and commit] I tend to disagree about using `GetAwaiter().GetResult()` instead of `async/await` + `ValueTask` (maybe it represent a gap in my knowledge). Even though it's done!

Why you disagree? I'd like to know!

ps: remember that in this case we are talking about something that **we know for sure** is already completed, see [here](https://github.com/ZiggyCreatures/FusionCache/pull/258/files#diff-c010c32326414424d5ebcf623c16a4ad42226bc47d7bd0a3853e664cb13b8c2cR378)

**Comment by jodydonetti** [comment]
Btw, regarding `.Result` vs `.GetAwaiter().GetResult()` see [here](https://devblogs.microsoft.com/pfxteam/asyncawait-faq/):

> Is there a functional difference between “task.Result” and “task.GetAwaiter().GetResult()”?
> 
> ... If the task ends in the RanToCompletion state, these are completely equivalent statements ...

So yeah they are "completely equivalent".

**Review by jodydonetti — APPROVED** [review]
LGTM

**Comment by henriqueholtz** [comment]
> Why you disagree? I'd like to know!
> 
> ps: remember that in this case we are talking about something that **we know for sure** is already completed, see [here](https://github.com/ZiggyCreatures/FusionCache/pull/258/files#diff-c010c32326414424d5ebcf623c16a4ad42226bc47d7bd0a3853e664cb13b8c2cR378)

@jodydonetti I see that the task is already finished. But according to the references bellow, `ValueTask` is useful for async operations which the result is already completed/computed (exactly the scenario right here, isn't?).

These advantages from `ValueTask` compared to a simple `Task` isn't enough to avoid your concern (repeated below)?

> An additional await means an additional async state machine and so on, which in this case (where we already know the task is completed) it's a waste. [...]

I can't understand why `GetAwaiter().GetResult()` is better than `ValueTask+async/await` in this scenario. Additional phrase which comes from references:

> ValueTask: It does not use any extra threads as a result. It also does not allocate an object on the managed heap.

Refs:
- [.NET Async Guide](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-taskfromresult-over-taskrun-for-pre-computed-or-trivially-computed-data)
- [Blog post "Understanding the Whys, Whats, and Whens of ValueTask"](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/)
- etc

**Comment by jodydonetti** [comment]
> @jodydonetti I see that the task is already finished. But according to the references bellow, `ValueTask` is useful for async operations which the result is already completed/computed

As [already mentioned](https://github.com/ZiggyCreatures/FusionCache/pull/258#issuecomment-2181471348):

> The task in question **is not usually already completed**, it's the (user provided) factory to grab the fresh version of the data, and it can go from being instantaneous to taking seconds or minutes, and we don't have control over that. I'm checking if, **at that moment**, is already completed but that **does not mean it usually is**.

Also, since what we are talking about is the **user-provided** factory, I can't just change the signature from `Task<T>` to `ValueTask<T>` without breaking all existing code out there, it's not doable.

> (exactly the scenario right here, isn't?)

Nope: the task we are talking about is the one for the **user-provided** factory, and that can take anything from a couple of milliseconds up to minutes, we don't know and we can't possibly know.

And we can't take a `Task`, which again can take anything from milliseconds to minutes, and if in one case it finishes quickly turn that into a `ValueTask` because it's better when "the operation already completed/computed", which again is not the case.

And we can't just change the signature to always be a `ValueTask` because changing the signature of the factory would be a breaking change.

To be clear: if I would create FusionCache from scratch today, I would have _probably_ specified the signature of the factory to use `ValueTask<TResult>` instead of `Task<TResult>`.

> These advantages from `ValueTask` compared to a simple `Task` isn't enough to avoid your concern (repeated below)?
> 
> > An additional await means an additional async state machine and so on, which in this case (where we already know the task is completed) it's a waste. [...]

No, because:
1) as said we can't change the signature of the factory from `Task` to `ValueTask`
2) here we can even skip the `await` completely, `Task` or `ValueTask` that is, since we already know the task is completed (but again, not "in general" and only after having checked that and on a case-by-case basis)

> Refs:
> - [.NET Async Guide](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-taskfromresult-over-taskrun-for-pre-computed-or-trivially-computed-data)

look at the header text:

> Prefer Task.FromResult over Task.Run for pre-computed or trivially computed data

in particular the final part

> for pre-computed or trivially computed data

As said, we don't know what the **user-provided** factory does: usually it is a call to a database, which is for sure non pre-computed, and 99.999% of the times non trivially compoted.

Hope this helps, anyway I'm merging this.

**Pull request merged**